software/contrib/hamlet.py

```diff
@@ -0,0 +1,359 @@
+                    self.track_multiplier = self.track_multiplier * 2
+                pass
+            else:
+                # Play previous CV Pattern, unless we are at the first pattern
+                if self.CvPattern != 0:
+                    self.CvPattern -= 1
+
+        # Triggered on each clock into digital input. Output triggers.
+        @din.handler
+        def clockTrigger():
+
+            self.step_length = len(self.BD[self.pattern])
+
+            # As the randomness value gets higher, the chance of a randomly selected int being lower gets higher
+            if randint(0,99) < self.randomness:
+                self.bd.value(randint(0, 1))
+                self.bd.value(randint(0, 1))
+            else:
+                # Trigger drums
+                self.bd.value(int(self.BD[self.pattern][self.drum_step]))
+                self.hh.value(int(self.HH[self.pattern][self.drum_step]))                    
+            
+            # A pattern was selected which is shorter than the current step. Set to zero to avoid an error
+            if self.drum_step >= self.step_length:
+                self.drum_step = 0 
+
+            track_1_cv, track_1_sparsity = self.track_1[self.CvPattern][self.track_step]
+            track_2_cv, track_2_sparsity = self.track_2[self.CvPattern][self.track_step]
+            
+            # Set track 1 and 2 voltage outputs based on previously
+            # generated random pattern and sparsity. CV only changed
+            # if sparsity value is above the current control level. So
+            # as sparsity increases, notes decrease.
+            
+            if track_1_sparsity > self.sparsity:
+                self.cv_1.voltage(track_1_cv)
+                self.gate_1.on()
+            else:
+                self.gate_1.off()
+                
+            if track_2_sparsity > self.sparsity:
+                self.cv_2.voltage(track_2_cv)
+                self.gate_2.on()
+            else:
+                self.gate_2.off()
+
+            # Reset clock step at 128 to avoid a HUGE integer if running for a long time
+            # over a really long period of time this would look like a memory leak
+            if self.clock_step < 128:
+                self.clock_step +=1
+            else:
+                self.clock_step = 0
+    
+            # Reset step number at step_length -1 as pattern arrays are zero-based
+            if self.drum_step < self.step_length - 1:
+                self.drum_step += 1
+            else:
+                self.drum_step = 0
+
+            if self.track_step < (self.step_length * self.track_multiplier) - 1:
+                self.track_step += 1
+            else:
+                self.track_step = 0
+
+                
+        @din.handler_falling
+        def clockTriggerEnd():
+            self.bd.off()
+            self.hh.off()
+            self.gate_1.off()
+            self.gate_2.off()
+
+    def generateNewRandomCVPattern(self):
```

Comment:
Optional [readability]: Consider adding docstring comments to these functions to help readers of this code.

https://peps.python.org/pep-0257/#one-line-docstrings